File: cses/09_geometry/1740_intersection-points/solution.py
import sys
from bisect import bisect_left, bisect_right


def main():
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    pos = 0
    n = int(data[pos]); pos += 1

    horizontals = []  # (x_left, x_right, y)
    verticals = []    # (x, y_low, y_high)
    ys = []           # tất cả y của đoạn ngang, dùng để nén tọa độ

    # Phân loại từng đoạn thành ngang (y1 == y2) hoặc dọc (x1 == x2).
    for _ in range(n):
        x1 = int(data[pos]); y1 = int(data[pos + 1])
        x2 = int(data[pos + 2]); y2 = int(data[pos + 3])
        pos += 4
        if y1 == y2:
            # đoạn ngang: theo ràng buộc x1 <= x2
            horizontals.append((x1, x2, y1))
            ys.append(y1)
        else:
            # đoạn dọc: x1 == x2, theo ràng buộc y1 <= y2
            verticals.append((x1, y1, y2))

    # Nén tọa độ y của các đoạn ngang thành rank.
    ys = sorted(set(ys))
    m = len(ys)

    # Tạo các sự kiện sweep line; type: add=0 < query=1 < remove=2 (thứ tự khi trùng x).
    # Nhờ bảo đảm không endpoint nào là giao điểm nên thứ tự này an toàn.
    events = []
    ap = events.append

    # Mỗi đoạn ngang -> sự kiện add tại xl và remove tại xr (rank 1-based cho BIT).
    for xl, xr, y in horizontals:
        r = bisect_left(ys, y) + 1
        ap((xl, 0, r, 0))
        ap((xr, 2, r, 0))

    # Mỗi đoạn dọc -> sự kiện query lấy khoảng rank [lo, hi] ứng với [y_low, y_high].
    for x, ylo, yhi in verticals:
        lo = bisect_left(ys, ylo)   # số phần tử < ylo
        hi = bisect_right(ys, yhi)  # số phần tử <= yhi
        if hi > lo:
            # range sum trên rank (lo+1 .. hi) = prefix(hi) - prefix(lo)
            ap((x, 1, lo, hi))

    events.sort()

    # BIT (Fenwick tree) trên trục y, kích thước m+1 (1-based).
    sz = m + 1
    tree = [0] * sz
    ans = 0

    # Quét theo x: add/remove cập nhật BIT (+1/-1), query cộng số đoạn ngang trong khoảng.
    for _x, typ, p, q in events:
        if typ == 1:
            # range sum = prefix(q) - prefix(p)
            s = 0
            i = q
            while i > 0:
                s += tree[i]
                i -= i & (-i)
            i = p
            while i > 0:
                s -= tree[i]
                i -= i & (-i)
            ans += s
        elif typ == 0:
            # point update +1: đoạn ngang bắt đầu hoạt động
            i = p
            while i < sz:
                tree[i] += 1
                i += i & (-i)
        else:  # typ == 2
            # point update -1: đoạn ngang kết thúc hoạt động
            i = p
            while i < sz:
                tree[i] -= 1
                i += i & (-i)

    sys.stdout.write(str(ans) + "\n")


main()

File: cses/09_geometry/1740_intersection-points/solution.cpp
#include <algorithm>
#include <iostream>
#include <tuple>
#include <vector>

// Sự kiện của sweep line: tại hoành độ x, loại type (add=0 < query=1 < remove=2).
// Với add/remove: left = rank của y trong BIT. Với query: [left, right) là khoảng rank.
struct Event {
    int x;
    int type;
    int left;
    int right;
};

// BIT (Fenwick tree) trên trục y: point update, prefix sum để đếm range động.
class FenwickTree {
public:
    explicit FenwickTree(int size)
        : size_(size), tree_(static_cast<std::size_t>(size) + 1U, 0) {}

    void add(int index, int value) {
        for (int i = index; i <= size_; i += i & -i) {
            tree_[static_cast<std::size_t>(i)] += value;
        }
    }

    int prefix_sum(int index) const {
        int result = 0;
        for (int i = index; i > 0; i -= i & -i) {
            result += tree_[static_cast<std::size_t>(i)];
        }
        return result;
    }

private:
    int size_;
    std::vector<int> tree_;
};

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n;
    std::cin >> n;

    struct Horizontal {
        int left_x;
        int right_x;
        int y;
    };
    struct Vertical {
        int x;
        int low_y;
        int high_y;
    };

    std::vector<Horizontal> horizontals;
    std::vector<Vertical> verticals;
    std::vector<int> horizontal_ys;  // các y của đoạn ngang, dùng để nén tọa độ
    horizontals.reserve(static_cast<std::size_t>(n));
    verticals.reserve(static_cast<std::size_t>(n));
    horizontal_ys.reserve(static_cast<std::size_t>(n));

    // Phân loại từng đoạn thành ngang (y1 == y2) hoặc dọc (x1 == x2).
    for (int i = 0; i < n; ++i) {
        int x1, y1, x2, y2;
        std::cin >> x1 >> y1 >> x2 >> y2;
        if (y1 == y2) {
            horizontals.push_back({x1, x2, y1});
            horizontal_ys.push_back(y1);
        } else {
            verticals.push_back({x1, y1, y2});
        }
    }

    // Nén tọa độ y của các đoạn ngang thành rank 1..m.
    std::sort(horizontal_ys.begin(), horizontal_ys.end());
    horizontal_ys.erase(
        std::unique(horizontal_ys.begin(), horizontal_ys.end()),
        horizontal_ys.end());

    std::vector<Event> events;
    events.reserve(2 * horizontals.size() + verticals.size());

    // Mỗi đoạn ngang -> sự kiện add tại xl và remove tại xr (rank 1-based).
    for (const Horizontal& segment : horizontals) {
        const int rank = static_cast<int>(std::lower_bound(
            horizontal_ys.begin(), horizontal_ys.end(), segment.y) -
            horizontal_ys.begin()) + 1;
        events.push_back({segment.left_x, 0, rank, 0});
        events.push_back({segment.right_x, 2, rank, 0});
    }

    // Mỗi đoạn dọc -> sự kiện query lấy khoảng rank [left, right) ứng với [low_y, high_y].
    for (const Vertical& segment : verticals) {
        const int left = static_cast<int>(std::lower_bound(
            horizontal_ys.begin(), horizontal_ys.end(), segment.low_y) -
            horizontal_ys.begin());
        const int right = static_cast<int>(std::upper_bound(
            horizontal_ys.begin(), horizontal_ys.end(), segment.high_y) -
            horizontal_ys.begin());
        if (left < right) {
            events.push_back({segment.x, 1, left, right});
        }
    }

    // Sắp xếp theo x tăng dần, trùng x thì theo type add < query < remove.
    std::sort(events.begin(), events.end(), [](const Event& a, const Event& b) {
        return std::tie(a.x, a.type) < std::tie(b.x, b.type);
    });

    // Quét: add/remove cập nhật BIT, query cộng số đoạn ngang đang hoạt động trong khoảng.
    FenwickTree active(static_cast<int>(horizontal_ys.size()));
    long long answer = 0;
    for (const Event& event : events) {
        if (event.type == 0) {
            active.add(event.left, 1);
        } else if (event.type == 1) {
            answer += active.prefix_sum(event.right) -
                      active.prefix_sum(event.left);
        } else {
            active.add(event.left, -1);
        }
    }

    std::cout << answer << '\n';
    return 0;
}
